content/docs/learn/typed-errors/validation.md

```diff
@@ -0,0 +1,312 @@
+approach, so we can give back as much information as possible to the user.
+This requires two changes to the code:
+- The result type is now a `NonEmptyList` of problems,
+- We need to wrap the different validations in `zipOrAccumulate`.
+
+<!--- INCLUDE
+sealed interface BookValidationError
+object EmptyTitle: BookValidationError
+object NoAuthors: BookValidationError
+data class EmptyAuthor(val index: Int): BookValidationError
+
+data class Author(val name: String)
+-->
+
+```kotlin
+data class Book private constructor(val title: String, val authors: List<Author>) {
+  companion object {
+    operator fun invoke(
+      title: String, authors: List<String>
+    ): Either<NonEmptyList<BookValidationError>, Book> = either {
+      zipOrAccumulate(
+        { ensure(title.isNotEmpty()) { EmptyTitle } },
+        { ensure(authors.isNotEmpty()) { NoAuthors } }
+      ) { _, _ ->
+        Book(title, TODO())
+      }
+    }
+  }
+}
+```
+<!--- KNIT example-validation-06.kt -->
+
+The result of each of the arguments of `zipOrAccumulate` are available in the
+trailing lambda. In this case we are not using them: we have the `title`
+already available, and for the list of authors we still need to perform the
+conversion from `String` to `Author`.
+
+:::tip Arguments to zipOrAccumulate
+
+All but the last argument to `zipOrAccumulate` represent the different
+validations we want to run aggregating their output. Those arguments are
+often _blocks_ wrapped in `{` curly braces `}`, which is a bit unusual for
+most Kotliners.
+
+:::
+
+## Validating a list
+
+The next step is turning the given `authors`, which is a list of `String`s, into
+a list of `Author`s. We need to run the smart constructor, thus, but at the same
+time we should accumulate any potential problems. Since this is related to
+author checks, we'll include this as part of the second validation.
+
+<!--- INCLUDE
+sealed interface BookValidationError
+object EmptyTitle: BookValidationError
+object NoAuthors: BookValidationError
+data class EmptyAuthor(val index: Int): BookValidationError
+
+object EmptyAuthorName
+
+data class Author private constructor(val name: String) {
+  companion object {
+    operator fun invoke(name: String): Either<EmptyAuthorName, Author> = either {
+      ensure(name.isNotEmpty()) { EmptyAuthorName }
+      Author(name)
+    }
+  }
+}
+-->
+
+```kotlin
+data class Book private constructor(val title: String, val authors: List<Author>) {
+  companion object {
+    operator fun invoke(
+      title: String, authors: List<String>
+    ): Either<NonEmptyList<BookValidationError>, Book> = either {
+      zipOrAccumulate(
+        { ensure(title.isNotEmpty()) { EmptyTitle } },
+        { 
+          ensure(authors.isNotEmpty()) { NoAuthors }
+          authors.withIndex().mapOrAccumulate {
+            Author(it.value)
+              .recover { _ -> raise(EmptyAuthor(it.index)) }
+              .bind()
+          }
```

Comment:
I think this code is incorrect, and is missing the `bind` call. Otherwise the syntax is `mapOrAccumulate(authors.withIndex()) { ... }`

```suggestion
          authors.withIndex().mapOrAccumulate {
            Author(it.value)
              .recover { _ -> raise(EmptyAuthor(it.index)) }
              .bind()
          }.bind()
```